internal/witness/cmd/client/main.go

```diff
@@ -0,0 +1,221 @@
+	wh "github.com/google/certificate-transparency-go/internal/witness/client/http"
+	"github.com/google/certificate-transparency-go/internal/witness/verifier"
+	"github.com/google/certificate-transparency-go/loglist3"
+)
+
+var (
+	logList   = flag.String("log_list_url", "https://www.gstatic.com/ct/log_list/v3/log_list.json", "The location of the log list")
+	witness   = flag.String("witness_url", "", "The endpoint of the witness HTTP API")
+	witnessPK = flag.String("witness_pk", "", "The base64-encoded witness public key")
+	interval  = flag.Duration("poll", 10*time.Second, "How frequently to poll to get new witnessed STHs")
+)
+
+// WitnessSigVerifier verifies the witness' signature on a cosigned STH.
+type WitnessSigVerifier interface {
+	VerifySignature(cosigned wit_api.CosignedSTH) error
+}
+
+// Witness consists of the witness' URL and signature verifier.
+type Witness struct {
+	Client   *wh.Witness
+	Verifier WitnessSigVerifier
+}
+
+type ctLog struct {
+	id       string
+	name     string
+	wsth     *wit_api.CosignedSTH
+	verifier *ct.SignatureVerifier
+}
+
+func main() {
+	flag.Parse()
+	ctx := context.Background()
+	// Set up the witness client.
+	wURL, err := url.Parse(*witness)
+	if err != nil {
+		glog.Exitf("Failed to parse witness URL: %v", err)
+	}
+	pk, err := ct.PublicKeyFromB64(*witnessPK)
+	if err != nil {
+		glog.Exitf("Failed to create witness public key: %v", err)
+	}
+	wv, err := verifier.NewWitnessVerifier(pk)
+	if err != nil {
+		glog.Exitf("Failed to create witness signature verifier: %v", err)
+	}
+	w := Witness{
+		Client: &wh.Witness{
+			URL: wURL,
+		},
+		Verifier: wv,
+	}
+	// Set up the log data.
+	ctLogs, err := populateLogs(*logList)
+	if err != nil {
+		glog.Exitf("Failed to set up log data: %v", err)
+	}
+	// Now poll the witness for each log.
+	wg := &sync.WaitGroup{}
+	for _, log := range ctLogs {
+		wg.Add(1)
+		go func(witness *Witness, log ctLog) {
+			defer wg.Done()
+			if err := log.getSTH(ctx, witness, *interval); err != nil {
+				glog.Errorf("getSTH: %v", err)
+			}
+		}(&w, log)
+	}
+	wg.Wait()
+}
+
+// populateLogs populates a list of ctLogs based on the log list.
+func populateLogs(logListURL string) ([]ctLog, error) {
+	u, err := url.Parse(logListURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URL: %v", err)
+	}
+	body, err := readURL(u)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get log list data: %v", err)
+	}
+	// Get data for all usable logs.
+	logList, err := loglist3.NewFromJSON(body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse JSON: %v", err)
+	}
+	usable := logList.SelectByStatus([]loglist3.LogStatus{loglist3.UsableLogStatus})
+	var logs []ctLog
+	for _, operator := range usable.Operators {
+		for _, log := range operator.Logs {
+			logID := base64.StdEncoding.EncodeToString(log.LogID)
+			logPK := base64.StdEncoding.EncodeToString(log.Key)
+			pk, err := ct.PublicKeyFromB64(logPK)
```

Comment:
`PublicKeyFromB64` just un-base64s and uses `x509.ParsePKIXPublicKey(der)` so you could avoid the re-de-encode and use the `ParsePKIX...` call directly here?

Reply:
Good call, done!